File: android_native/app/src/main/java/com/familymotion/app/data/model/PoseData.kt
package com.familymotion.app.data.model

import com.google.mlkit.vision.pose.Pose
import com.google.mlkit.vision.pose.PoseLandmark

/**
 * 骨骼关键点数据
 */
data class PoseData(
    val landmarks: List<PoseLandmarkData>,
    val timestamp: Long = System.currentTimeMillis()
) {
    companion object {
        fun fromPose(pose: Pose): PoseData {
            val landmarks = pose.allPoseLandmarks.map { landmark ->
                PoseLandmarkData(
                    type = mapLandmarkType(landmark.landmarkType),
                    x = landmark.position.x,
                    y = landmark.position.y,
                    z = landmark.position3D.z,
                    inFrameLikelihood = landmark.inFrameLikelihood
                )
            }
            return PoseData(landmarks = landmarks)
        }

        private fun mapLandmarkType(type: Int): LandmarkType {
            return when (type) {
                PoseLandmark.NOSE -> LandmarkType.NOSE
                PoseLandmark.LEFT_EYE_INNER -> LandmarkType.LEFT_EYE_INNER
                PoseLandmark.LEFT_EYE -> LandmarkType.LEFT_EYE
                PoseLandmark.LEFT_EYE_OUTER -> LandmarkType.LEFT_EYE_OUTER
                PoseLandmark.RIGHT_EYE_INNER -> LandmarkType.RIGHT_EYE_INNER
                PoseLandmark.RIGHT_EYE -> LandmarkType.RIGHT_EYE
                PoseLandmark.RIGHT_EYE_OUTER -> LandmarkType.RIGHT_EYE_OUTER
                PoseLandmark.LEFT_EAR -> LandmarkType.LEFT_EAR
                PoseLandmark.RIGHT_EAR -> LandmarkType.RIGHT_EAR
                PoseLandmark.LEFT_MOUTH -> LandmarkType.LEFT_MOUTH
                PoseLandmark.RIGHT_MOUTH -> LandmarkType.RIGHT_MOUTH
                PoseLandmark.LEFT_SHOULDER -> LandmarkType.LEFT_SHOULDER
                PoseLandmark.RIGHT_SHOULDER -> LandmarkType.RIGHT_SHOULDER
                PoseLandmark.LEFT_ELBOW -> LandmarkType.LEFT_ELBOW
                PoseLandmark.RIGHT_ELBOW -> LandmarkType.RIGHT_ELBOW
                PoseLandmark.LEFT_WRIST -> LandmarkType.LEFT_WRIST
                PoseLandmark.RIGHT_WRIST -> LandmarkType.RIGHT_WRIST
                PoseLandmark.LEFT_PINKY -> LandmarkType.LEFT_PINKY
                PoseLandmark.RIGHT_PINKY -> LandmarkType.RIGHT_PINKY
                PoseLandmark.LEFT_INDEX -> LandmarkType.LEFT_INDEX
                PoseLandmark.RIGHT_INDEX -> LandmarkType.RIGHT_INDEX
                PoseLandmark.LEFT_THUMB -> LandmarkType.LEFT_THUMB
                PoseLandmark.RIGHT_THUMB -> LandmarkType.RIGHT_THUMB
                PoseLandmark.LEFT_HIP -> LandmarkType.LEFT_HIP
                PoseLandmark.RIGHT_HIP -> LandmarkType.RIGHT_HIP
                PoseLandmark.LEFT_KNEE -> LandmarkType.LEFT_KNEE
                PoseLandmark.RIGHT_KNEE -> LandmarkType.RIGHT_KNEE
                PoseLandmark.LEFT_ANKLE -> LandmarkType.LEFT_ANKLE
                PoseLandmark.RIGHT_ANKLE -> LandmarkType.RIGHT_ANKLE
                PoseLandmark.LEFT_HEEL -> LandmarkType.LEFT_HEEL
                PoseLandmark.RIGHT_HEEL -> LandmarkType.RIGHT_HEEL
                PoseLandmark.LEFT_FOOT_INDEX -> LandmarkType.LEFT_FOOT_INDEX
                PoseLandmark.RIGHT_FOOT_INDEX -> LandmarkType.RIGHT_FOOT_INDEX
                else -> LandmarkType.NOSE
            }
        }
    }
}

/**
 * 单个骨骼关键点
 */
data class PoseLandmarkData(
    val type: LandmarkType,
    val x: Float,
    val y: Float,
    val z: Float = 0f,
    val inFrameLikelihood: Float = 0f
)

/**
 * ML Kit 骨骼关键点类型映射
 */
enum class LandmarkType {
    NOSE,
    LEFT_EYE_INNER, LEFT_EYE, LEFT_EYE_OUTER,
    RIGHT_EYE_INNER, RIGHT_EYE, RIGHT_EYE_OUTER,
    LEFT_EAR, RIGHT_EAR,
    LEFT_MOUTH, RIGHT_MOUTH,
    LEFT_SHOULDER, RIGHT_SHOULDER,
    LEFT_ELBOW, RIGHT_ELBOW,
    LEFT_WRIST, RIGHT_WRIST,
    LEFT_PINKY, RIGHT_PINKY,
    LEFT_INDEX, RIGHT_INDEX,
    LEFT_THUMB, RIGHT_THUMB,
    LEFT_HIP, RIGHT_HIP,
    LEFT_KNEE, RIGHT_KNEE,
    LEFT_ANKLE, RIGHT_ANKLE,
    LEFT_HEEL, RIGHT_HEEL,
    LEFT_FOOT_INDEX, RIGHT_FOOT_INDEX
}

/**
 * 骨骼关键点索引常量
 */
object LandmarkIndices {
    const val NOSE = 0
    const val LEFT_SHOULDER = 11
    const val RIGHT_SHOULDER = 12
    const val LEFT_ELBOW = 13
    const val RIGHT_ELBOW = 14
    const val LEFT_WRIST = 15
    const val RIGHT_WRIST = 16
    const val LEFT_HIP = 23
    const val RIGHT_HIP = 24
    const val LEFT_KNEE = 25
    const val RIGHT_KNEE = 26
    const val LEFT_ANKLE = 27
    const val RIGHT_ANKLE = 28
}

File: ios_native/FamilyMotion/Models/PoseData.swift
//
//  PoseData.swift
//  FamilyMotion
//
//  Created by FamilyMotion on 2026/06/10.
//

import Foundation
import CoreGraphics
import Vision

// MARK: - PoseData 骨骼数据
struct PoseData: Identifiable, Codable, Hashable {
    var id: UUID = UUID()
    var landmarks: [PoseLandmark] = []
    var timestamp: Date = Date()
    var confidence: Float = 0.0

    func landmark(for type: LandmarkType) -> PoseLandmark? {
        return landmarks.first { $0.type == type }
    }
}

// MARK: - PoseLandmark 单个骨骼点
struct PoseLandmark: Codable, Hashable {
    var type: LandmarkType
    var x: Float
    var y: Float
    var z: Float = 0.0
    var confidence: Float = 0.0

    var point: CGPoint {
        return CGPoint(x: CGFloat(x), y: CGFloat(y))
    }
}

// MARK: - LandmarkType 骨骼点类型
enum LandmarkType: String, Codable, CaseIterable, Hashable {
    case nose = "鼻子"
    case leftEyeInner = "左眼内侧"
    case leftEye = "左眼"
    case leftEyeOuter = "左眼外侧"
    case rightEyeInner = "右眼内侧"
    case rightEye = "右眼"
    case rightEyeOuter = "右眼外侧"
    case leftEar = "左耳"
    case rightEar = "右耳"
    case leftMouth = "左嘴角"
    case rightMouth = "右嘴角"
    case leftShoulder = "左肩"
    case rightShoulder = "右肩"
    case leftElbow = "左肘"
    case rightElbow = "右肘"
    case leftWrist = "左腕"
    case rightWrist = "右腕"
    case leftPinky = "左小指"
    case rightPinky = "右小指"
    case leftIndex = "左食指"
    case rightIndex = "右食指"
    case leftThumb = "左拇指"
    case rightThumb = "右拇指"
    case leftHip = "左髋"
    case rightHip = "右髋"
    case leftKnee = "左膝"
    case rightKnee = "右膝"
    case leftAnkle = "左踝"
    case rightAnkle = "右踝"
    case leftHeel = "左脚跟"
    case rightHeel = "右脚跟"
    case leftFootIndex = "左脚尖"
    case rightFootIndex = "右脚尖"
}

// MARK: - PoseData 的工厂方法
extension PoseData {

    /// 从 Apple Vision 的 VNHumanBodyPoseObservation 转换为 PoseData
    static func fromHumanBodyPose(
        _ observation: VNHumanBodyPoseObservation,
        imageSize: CGSize
    ) -> PoseData? {

        var landmarks: [PoseLandmark] = []
        var totalConfidence: Float = 0.0
        var detectedCount = 0

        // 定义关节点映射（VNPointKey 到 LandmarkType）
        let jointMapping: [VNHumanBodyPoseObservation.JointName: LandmarkType] = [
            .nose: .nose,
            .leftEye: .leftEye,
            .rightEye: .rightEye,
            .leftEar: .leftEar,
            .rightEar: .rightEar,
            .leftShoulder: .leftShoulder,
            .rightShoulder: .rightShoulder,
            .leftElbow: .leftElbow,
            .rightElbow: .rightElbow,
            .leftWrist: .leftWrist,
            .rightWrist: .rightWrist,
            .leftHip: .leftHip,
            .rightHip: .rightHip,
            .leftKnee: .leftKnee,
            .rightKnee: .rightKnee,
            .leftAnkle: .leftAnkle,
            .rightAnkle: .rightAnkle
        ]

        for (jointName, landmarkType) in jointMapping {
            do {
                let point = try observation.recognizedPoint(jointName)
                if point.confidence > 0.3 {
                    // Vision 的坐标是归一化的 (0-1)，并且 y 轴从上到下
                    // 这里需要翻转 y 以匹配常规屏幕坐标
                    let x = point.location.x
                    let y = 1.0 - point.location.y

                    let landmark = PoseLandmark(
                        type: landmarkType,
                        x: Float(x),
                        y: Float(y),
                        z: 0.0,
                        confidence: Float(point.confidence)
                    )
                    landmarks.append(landmark)
                    totalConfidence += Float(point.confidence)
                    detectedCount += 1
                }
            } catch {
                continue
            }
        }

        guard detectedCount > 0 else { return nil }

        let averageConfidence = totalConfidence / Float(detectedCount)

        return PoseData(
            landmarks: landmarks,
            timestamp: Date(),
            confidence: averageConfidence
        )
    }
}
